finality-aleph/src/sync/service.rs

```diff
@@ -0,0 +1,274 @@
+    }
+
+    fn delayed_request(&mut self, block_id: BlockIdFor<J>) {
+        self.tasks.schedule_in(block_id, Duration::from_millis(500));
+    }
+
+    fn request(&mut self, block_id: BlockIdFor<J>) {
+        self.tasks.schedule_in(block_id, Duration::ZERO);
+    }
+
+    fn broadcast(&mut self) {
+        let state = match self.handler.state() {
+            Ok(state) => state,
+            Err(e) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "Failed to construct own knowledge state: {}.", e
+                );
+                return;
+            }
+        };
+        let data = NetworkData::StateBroadcast(state);
+        if let Err(e) = self.network.broadcast(data) {
+            warn!(target: LOG_TARGET, "Error sending broadcast: {}.", e);
+        }
+    }
+
+    fn send_request_for(
+        &mut self,
+        block_id: BlockIdFor<J>,
+        branch_knowledge: BranchKnowledge<J>,
+        peers: HashSet<N::PeerId>,
+    ) {
+        let state = match self.handler.state() {
+            Ok(state) => state,
+            Err(e) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "Failed to construct own knowledge state: {}.", e
+                );
+                return;
+            }
+        };
+        let request = Request::new(block_id, branch_knowledge, state);
+        let data = NetworkData::Request(request);
+        if let Err(e) = self.network.send_to_random(data, peers) {
+            warn!(target: LOG_TARGET, "Error sending request: {}.", e);
+        }
+    }
+
+    fn send_to(&mut self, data: NetworkData<J>, peer: N::PeerId) {
+        if let Err(e) = self.network.send_to(data, peer) {
+            warn!(target: LOG_TARGET, "Error sending response: {}.", e);
+        }
+    }
+
+    fn perform_sync_action(&mut self, action: SyncAction<J>, peer: N::PeerId) {
+        use SyncAction::*;
+        match action {
+            Response(data) => self.send_to(data, peer),
+            Task(block_id) => self.request(block_id),
+            Noop => (),
+        }
+    }
+
+    fn handle_state(&mut self, state: State<J>, peer: N::PeerId) {
+        match self.handler.handle_state(state, peer.clone()) {
+            Ok(action) => self.perform_sync_action(action, peer),
+            Err(e) => warn!(
+                target: LOG_TARGET,
+                "Error handling sync state from {:?}: {}.", peer, e
+            ),
+        }
+    }
+
+    fn handle_justifications(&mut self, justifications: Vec<J::Unverified>, peer: N::PeerId) {
+        use SyncAction::*;
+        let mut previous_block_id = None;
+        for justification in justifications {
+            let maybe_block_id = match self
+                .handler
+                .handle_justification(justification, peer.clone())
+            {
+                Ok(action) => match action {
+                    Response(data) => {
```

Comment:
when will this branch run?

Reply:
Hm, right, reworked the interface a tad.